Switched RulesOfHooks.js to use BigInt. Added test and updated .eslin…

scyron6 opened the pull request:
…trc.js to use es2020.

<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `main`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test --prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

<!--
-->
Fixes #24279. JavaScript could not handle the large integer when there were numerous conditional statements before and after the hook. This was causing rounding errors which resulted in a conditional being evaluated incorrectly. I switched the functions to use BigInts instead and updated the eslintrc.js. If using BigInts is not acceptable, I'll be happy to try and find a different solution. 

## How did you test this change?

<!--
  Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface.
  How exactly did you verify that your PR solves the issue you wanted to solve?
  If you leave this empty, your PR will very likely be closed.
-->
The user who created the issue linked to a repository that showed the bug. I copied that exact code and pasted into the tests file. This test now passes.
Link to repository: https://github.com/SanderRonde/eslint-hook-bug

Now when I open that file in my local instance, there is no error message for the hook.
![image](https://user-images.githubusercontent.com/63068182/162086280-bae69c58-2285-4813-a10a-9c0d03a8a770.png)

When I run all tests in ESLintRulesOfHooks-test.js, everything passes.
![image](https://user-images.githubusercontent.com/63068182/162086653-d428aec6-b953-4d7b-9020-7647259ff629.png)

Unfortunately, when I ran yarn test --prod, I got several errors that were unrelated to ESLintRulesOfHooks. I am not sure if this is normal.

gaearon commented:
Fascinating.

gaearon commented:
It seems worrying that we get to such high numbers. Does this also mean we're doing a ton of work and are slow in this case? Is this case pathological in some particular way? Can we think of any way to do less work for it?

gaearon commented:
cc @calebmer you might find this interesting

scyron6 commented:
### Work Context
Adding some screenshots from when I was working just to add some context. This first screenshot is before I added BigInts.  The pathsFromStartToEnd and allPathsFromStartToEnd are the values that are being compared in the conditional for the error.
![image](https://user-images.githubusercontent.com/63068182/162190795-b78e0be0-570c-4185-b71c-dcf7a04d3389.png)

These are the values from the same code but with BigInts. 
![image](https://user-images.githubusercontent.com/63068182/162190993-686ad60f-d79d-4a07-97a6-f923a736f5ff.png)

The code is using a recursive algorithm with time O(log n). I added a console log to the function to see the value at each step of recursion for countPathsFromStart. 
![image](https://user-images.githubusercontent.com/63068182/162191370-b81ef274-86e8-414b-89e4-b7eaa6cd2b9d.png)

### Concluding Thoughts
I do not see any way to get around having very large integers if a user has complex code with a lot of conditionals. I have not done a deep dive into the recursive algorithm, but it seems well designed and runs in O(log n). As we have to make sure the hook is reachable from the initial segment and every final segment, I believe we will have to traverse the code somehow. 

I'm wondering if there is a way to do this with simple multiplication. If there are 4 simple if else statements before a hook, then there are 2^4 pathsFromStart. We'd have to distinguish the different types of conditionals somehow, which the current recursive algorithm already takes care of.

gaearon commented:
Can you do some kind of a stress test to see how this change affects performance for the common cases? I'm not sure if there are any perf pitfalls to using BigInts.

scyron6 commented:
### Setup
I took the eslint-hook-bug repository and made some adjustments. I copied app/foo.tsx (the file that previously failed because the ints were too big) and copied it to fail.tsx. I then removed a couple of the conditionals from app/foo.tsx so there would be no errors when run with the current version. Finally, I copied app/foo.tsx 100 times into the app folder for the stress test.

### Stress Test
I checked out the main branch and ran yarn build eslint-plugin-react-hooks to simulate the current environment.
![image](https://user-images.githubusercontent.com/63068182/162284451-576c5371-1ff9-47a4-9432-6458f6494a66.png)

Eslint comes with feature that shows you how long each run takes. The command 'TIMING=1 yarn eslint app/*' will run the rule on every file in app and give us the total time eslint took to run. I ran the command 10 times and sent the output to main_runs.txt.  I then ran the same command but with fail.tsx and also sent that output to main_runs.txt just so we can see the failure.
[main_runs.txt](https://github.com/facebook/react/files/8446148/main_runs.txt)


I then switched over to the branch that with BigInts and reran yarn build. 
![image](https://user-images.githubusercontent.com/63068182/162285824-6017f1b7-bb7f-4412-a644-f0816e213b15.png)

I then ran the same TIMING=1 yarn eslint app/* command 10 times and sent the output to big_int_runs.txt. I then ran it with fail.tsx to watch it pass this time. 
[big_int_runs.txt](https://github.com/facebook/react/files/8446150/big_int_runs.txt)


### Concluding Thoughts
The runtimes seemed very similar to me. If you would also like me to check the memory usage difference, let me know. That might take longer though as I'd have to do a lot of research on how to do that.

gaearon commented:
Thanks for checking. Would you mind fixing the lint issue? I think you need to find the lint config we use for builds and add BigInt to known globals.

gaearon commented on yarn.lock:
```diff
@@ -13629,7 +13629,7 @@ rc@^1.0.1, rc@^1.1.6, rc@^1.2.8:
     prop-types "^15.6.2"
     scheduler "^0.13.0"
 
-react-is@^16.8.1, "react-is@npm:react-is":
```
Let’s exclude this change?

scyron6 replied:
Yep got rid of it.

gaearon commented on .eslintrc.js:
```diff
@@ -15,6 +15,9 @@ module.exports = {
   // Stop ESLint from looking for a configuration file in parent folders
   root: true,
+  env: {
+    es2020: 'true',
```
What are the implications of this setting? Are we going to accidentally allow more things in other source files which we aren’t meant to use (because we aim to be compatible with older browsers)?

scyron6 replied:
You're right there could be many more implications that I didn't consider. I got rid of that change and added the comment /* global BigInt */ to the RulesOfHooks.js to get rid of the eslint error.

gaearon commented:
Looks like your commits aren't attributed to your GH user (maybe different email). Would you like to fix them up so that you appear as a contributor after merge? I'm not sure how this works.

scyron6 commented:
You were right I had an old email in there. Don't worry about it though.
